### PR TITLE
OFS-134: 'update contract' service with some perms in Contract

### PR DESCRIPTION
**Pull request by sniedzielski**
In that PR I want to merge 'update contract' service on backend side according to the wiki page https://openimis.atlassian.net/wiki/spaces/OP/pages/1730740273/Contract+Module

Including some changes related to the changes in Core module with dates (migration file with changing date type of fields into core one instead of django default one). https://github.com/openimis/openimis-be-core_py/pull/36

I made this as a "Draft" PR.

### CODE REVIEW

File: contract/services.py [hunk 1]
```diff
@@ -75,15 +87,71 @@ def __evaluate_contract_valuation(self, contract_details_result):
         )
         return result_contract_valuation["data"]["total_amount"]
 
+    # TODO update contract scenario according to wiki page
+    @check_authentication
+    def update(self, contract):
+        try:
+            # check rights for contract / amendments
```
Comment (sniedzielski) [inline contract/services.py:94]: check rights for contract / amendments

File: contract/services.py [hunk 2]
```diff
@@ -75,15 +87,71 @@ def __evaluate_contract_valuation(self, contract_details_result):
...
+            if not (self.user.has_perms(ContractConfig.gql_mutation_update_contract_perms) or self.user.has_perms(
+                    ContractConfig.gql_mutation_approve_ask_for_change_contract_perms)):
+                raise PermissionError("Unauthorized")
+            updated_contract = ContractModel.objects.filter(id=contract['id']).first()
+            # updatable scenario
```
Comment (sniedzielski) [inline contract/services.py:99]: check status: if in “draft”/ “request for information“ / “counter“ only the  code cannot be changed,

File: contract/services.py [hunk 3]
```diff
@@ -75,15 +87,71 @@ def __evaluate_contract_valuation(self, contract_details_result):
...
+            if self.__check_rights_by_status(updated_contract.state) == "updatable":
+                if "code" in contract:
+                    raise ContractUpdateError("That fields are not editable in that permission")
+                return _output_result_success(
+                    dict_representation=self.__update_contract_fields(
+                        contract_input=contract,
+                        updated_contract=updated_contract
+                    )
+                )
+            # approvable scenario
```
Comment (sniedzielski) [inline contract/services.py:109]: check satus: if in “Negotiable” changes are possible only with the authority “Approve/ask for change”

File: contract/services.py [hunk 4]
```diff
@@ -75,15 +87,71 @@ def __evaluate_contract_valuation(self, contract_details_result):
...
+                return _output_result_success(
+                    dict_representation=self.__update_contract_fields(
+                        contract_input=contract,
+                        updated_contract=updated_contract
+                    )
+                )
+        except Exception as exc:
+            return _output_exception(model_name="ContractModule", method="update", exception=exc)
+
+    def __check_rights_by_status(self, status):
```
Comment (sniedzielski) [inline contract/services.py:123]: private helper function to check rights by status etc

File: contract/services.py [hunk 5]
```diff
@@ -75,15 +87,71 @@ def __evaluate_contract_valuation(self, contract_details_result):
...
+        state = "cannot_update"
+        if status in [ContractModel.STATE_DRAFT, ContractModel.STATE_REQUEST_FOR_INFORMATION, ContractModel.STATE_COUNTER]:
+           state = "updatable"
+        if status == ContractModel.STATE_NEGOTIABLE:
+            state = "approvable"
+        return state
+
+    def __update_contract_fields(self, contract_input, updated_contract):
```
Comment (sniedzielski) [inline contract/services.py:131]: private helper function to update fields in Contract

File: contract/migrations/0009_auto_20210118_1427.py [hunk 6]
```diff
@@ -0,0 +1,95 @@
+# Generated by Django 3.0.3 on 2021-01-18 14:27
+
```
Comment (sniedzielski) [inline contract/migrations/0009_auto_20210118_1427.py:2]: migration file related to the change in https://github.com/openimis/openimis-be-core_py/pull/36

File: contract/apps.py [hunk 7]
```diff
@@ -1,5 +1,43 @@
 from django.apps import AppConfig
```
Comment (sniedzielski) [inline contract/apps.py:3]: add some perms, later I will add another ones in "OFS-154: create contract role for admin"

File: contract/apps.py [hunk 8]
```diff
@@ -1,5 +1,43 @@
...
+MODULE_NAME = "contract"
+
+
+DEFAULT_CFG = {
+    "gql_query_contract_perms": ["152101"],
+    "gql_query_contract_admins_perms": [],
+    "gql_mutation_create_contract_perms": ["152102"],
+    "gql_mutation_update_contract_perms": ["152103"],
+    "gql_mutation_approve_ask_for_change_contract_perms": ["152108"],
+}
```
Comment (delcroip) [inline contract/apps.py:13]: we might have to rework that part later to wrap it in a "perms" dict add some label for the Profile page 

"perms":{"contract_perms": {"id":"152101", "label":{"fr":"Contrat", "en":"Contract"}}} ....

**Review by delcroip — APPROVED** [review]

File: contract/apps.py (hunk 8 above)
Comment (sniedzielski) [reply contract/apps.py:13]: Ok. We can discuss this later related to the new Profile page etc.

**Pull request merged**